### PR TITLE
Enh/symmetry adapted basis

### PR DESCRIPTION
**Pull request by kalmarek**

### CODE REVIEW

**Comment by kalmarek** [comment]
We seriously need to increase the coverage ;)

**Review by tweisser — APPROVED** [review]

File: examples/action_polynomials.jl [hunk 1]
```diff
@@ -0,0 +1,16 @@
+using DynamicPolynomials
+using MultivariatePolynomials
+
+function SymbolicWedderburn.InducingHomomorphism(basis::Union{<:MonomialVector, AbstractVector{<:Monomial}})
```
Comment (blegat) [inline examples/action_polynomials.jl:4]: You can do `AbstractVector{<:AbstractMonomial}`

File: src/projections.jl [hunk 2]
```diff
@@ -24,29 +41,120 @@ function matrix_projection(
...
+
+The coefficients of the invariant basis are returned in (orthogonal) blocks
+corresponding to irreducible characters of `G`.
+
+!!! Note:
+Each block is invariant under the action of `G`, i.e. the action may permute
+vectors from symmetry adapted basis within each block.
+"""
+function symmetry_adapted_basis_float(G::PermutationGroups.PermGroup, basis)
+    chars = characters_dixon(G)
```
Comment (blegat) [inline src/projections.jl:141]: You are missing a line `ihom = InducingHomomorphism(basis)`

**Review by blegat — APPROVED** [review]

**Comment by kalmarek** [comment]
mein Gott, when broadcasting is fixed on master, now master decided to change spacing in parametrized types...

**Comment by blegat** [comment]
Anything blocking the merge ?

**Comment by kalmarek** [comment]
@blegat: coverage ;)
@tweisser should I pick it up?

**Comment by kalmarek** [comment]
@blegat I will add coverage and merge this over this weekend

**Comment by kalmarek** [comment]
@tweisser @blegat this seems to be ready to go!

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -10,14 +10,17 @@ jobs:
       matrix:
         version:
           - '1.0'
-          - '1.4'
+          - '1.5'
```
Comment (blegat) [inline .github/workflows/ci.yml:13]: You can use `'1'` to target the latest `1.x`

File: Project.toml [hunk 4]
```diff
@@ -1,17 +1,22 @@
 name = "SymbolicWedderburn"
 uuid = "858aa9a9-4c7c-4c62-b466-2421203962a2"
 authors = ["Marek Kaluba <kalmar@amu.edu.pl>", "tweisser <tillmann.weisser@web.de>"]
-version = "0.1.0"
+version = "0.2.0"
```
Comment (blegat) [inline Project.toml:4]: Why updating the bound ? This isn't released yet

File: examples/ex_C4.jl [hunk 5]
```diff
@@ -12,86 +12,38 @@ let m = SOSModel(SCS.Optimizer)
     @variable m t
     @objective m Max t
     @variable m sos SOSPoly([1; x])
-    @constraint m sum(x) + sum(x.^2) - t == sos
+    @constraint m sum(x) + sum(x .^ 2) - t == sos
```
Comment (blegat) [inline examples/ex_C4.jl:15]: `^` is an exception to the rule that there should be space around binary operators:
https://jump.dev/JuMP.jl/v0.21.6/style/#Exceptions

File: src/characters.jl [hunk 6]
```diff
@@ -40,59 +40,60 @@ if VERSION >= v"1.3.0"
...
     function (χ::VirtualCharacter)(g::PermutationGroups.GroupElem)
         for (i, cc) in enumerate(conjugacy_classes(χ))
             g ∈ cc && return χ[i]
         end
-        throw(DomainError(
-            g,
-            "element does not belong to conjugacy classes of χ",
-        ))
+        throw(
+            DomainError(g, "element does not belong to conjugacy classes of χ"),
```
Comment (blegat) [inline src/characters.jl:61]: `χ` or `$χ` ?

**Comment by kalmarek** [comment]
@blegat I'm updating the examples and I tried this:

```julia
msym = let m = SOSModel(SCS.Optimizer), basis = basis, R = R
    @variable m t
    @objective m Max t
    soses = [@variable m [1:length(R)] SOSPoly(FixedPolynomialBasis(r * basis)) for r in R]
    @constraint m f - t == sum(soses)

    optimize!(m)
    @info termination_status(m) objective_value(m)
    msym
end
```
but JuMP/SumOfSquares refuses to `sum(soses)`; any ideas?

**Comment by blegat** [comment]
This looks to be a but in MutableArithmetics, I would recommend doing the sum outside the macro as a workaround.

**Comment by kalmarek** [comment]
> This looks to be a but in MutableArithmetics, I would recommend doing the sum outside the macro as a workaround.
> **EDIT** in fact, this should be resolved by doing
> 
> ```julia
> soses = @variable m [1:length(R)] SOSPoly(FixedPolynomialBasis(r * basis))
> ```
> 
> Otherwise, you're doing a vector of vector and `sum` gives a vector instead of a polynomial

??? this doesn't even compile as `r` is not defined. did you mean
```julia
soses = @variable m [1:length(R)] SOSPoly(FixedPolynomialBasis(R * basis))
```
?? (non-compatible dimensions, doesn't work even with `R .* Ref(basis)`)
this: `polys = SOSPoly.(FixedPolynomialBasis.(R.*Ref(basis)))` gives me a vector of SOSPolys. How do I create a vector of SOSPolys variables ??

**Comment by blegat** [comment]
`[@variable m variable_type=SOSPoly(FixedPolynomialBasis(r * basis)) for r in R]`

File: examples/ex_C4.jl (hunk 5 above)
Comment (kalmarek) [reply examples/ex_C4.jl]: as long as this is not part of JuMP, I'd rather rely on automatic formatting for those details.

**Review by blegat — APPROVED** [review]

**Review by tweisser — APPROVED** [review]

**Comment by kalmarek** [comment]
:tada:

**Pull request merged**